rust/sbp/src/lib.rs

```diff
@@ -22,66 +22,157 @@ pub type Result<T> = result::Result<T, Error>;
 
 pub const SBP_MAX_PAYLOAD_SIZE: usize = 255;
 
+pub trait SbpString {
+    fn as_bytes(&self) -> &[u8];
+    fn to_string(&self) -> String;
+}
+
 #[derive(Debug, Clone)]
-pub struct SbpString(Vec<u8>);
+pub struct UnboundedSbpString(Vec<u8>);
 
-impl SbpString {
-    pub fn as_bytes(&self) -> &[u8] {
+impl SbpString for UnboundedSbpString {
+    fn as_bytes(&self) -> &[u8] {
         &self.0
     }
-    pub fn to_string(&self) -> String {
+    fn to_string(&self) -> String {
         String::from_utf8_lossy(&self.0).into()
     }
 }
 
+impl fmt::Display for UnboundedSbpString {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "SbpString({})", SbpString::to_string(self))
+    }
+}
+
+#[cfg(feature = "sbp_serde")]
+impl<'de> Deserialize<'de> for UnboundedSbpString {
+    fn deserialize<D>(deserializer: D) -> result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Deserialize::deserialize(deserializer).map(|s: String| UnboundedSbpString::from(s))
+    }
+}
+
 #[cfg(feature = "sbp_serde")]
-impl Serialize for SbpString {
+impl Serialize for UnboundedSbpString {
     fn serialize<S>(&self, serializer: S) -> result::Result<S::Ok, S::Error>
     where
         S: Serializer,
     {
-        let s: String = self.clone().into();
+        let s = SbpString::to_string(self);
         serializer.serialize_str(&s)
     }
 }
 
+impl From<UnboundedSbpString> for String {
+    fn from(s: UnboundedSbpString) -> String {
+        SbpString::to_string(&s)
+    }
+}
+
+impl From<UnboundedSbpString> for Vec<u8> {
+    fn from(s: UnboundedSbpString) -> Vec<u8> {
+        s.0
+    }
+}
+
+impl From<String> for UnboundedSbpString {
+    fn from(s: String) -> UnboundedSbpString {
+        UnboundedSbpString(s.into_bytes())
+    }
+}
+
+impl TryFrom<&mut &[u8]> for UnboundedSbpString {
+    type Error = Error;
+    fn try_from(buf: &mut &[u8]) -> Result<UnboundedSbpString> {
+        let amount = buf.len();
+        let (head, tail) = buf.split_at(amount);
+        *buf = tail;
+        Ok(UnboundedSbpString(head.to_vec()))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct BoundedSbpString<const SIZE: usize>([u8; SIZE]);
```

Comment:
This is the only intentional change in functionality, when the SBP message specifies a fixed length string we store the entire string instead of up to the maximum length.